### DISCUSSION TITLE
The IAM wild wildcard in its field of resource doesn't work

### DISCUSSION THREAD

**Discussion by bangbangOk** [seaweedfs/seaweedfs#4233] [Q&A]
Here is my IAM policy document,
{
	"Statement": [{
		"Action": ["s3:Get*"],
		"Effect": "Allow",
		"Resource": ["arn:aws:s3:::mybucket/test/*"]
	}],
	"Version": "2012-10-17"
}

I use AWS S3 Java SDK to put user policy.

The API invocation is successful. But when I checked the s3.configure by running the command echo 's3.configure' | weed shell, the identity I saw is incorrect, the “actions” of which is an empty array.
{
      "name":  "test-user",
      "credentials":  [
        {
          "accessKey":  "abcdefg",
          "secretKey":  "abcedfg1"
        }
      ],
      "actions":  [],
      "accountId":  ""
}

Any help is appreciated!

**Comment by chrislusf** [comment]
How to reproduce this? Please share the exact steps.

**Reply by bangbangOk** [reply]
Sure. I use the official AWS IAM Java SDK(aws-java-sdk-iam-1.11.313) to talk with SeaweedFS IAM service.

I have created a bucket, and uploaded a file through seaweedfs web UI to this bucket.
/buckets/mybucket/myfolder/avatar.png

I want to create an identity which has Read permission to access any file with the key matching the format "/mybucket/myfolder/*"

 Here are my steps:

step-1：
I created an IAM user.

final CreateUserResult createUserResult = iam.createUser(new CreateUserRequest().withUserName("testUser"));

step-2：
I created an accessKey and associated it with the IAM user.

final CreateAccessKeyResult createAccessKeyResult =
                iam.createAccessKey(new CreateAccessKeyRequest().withUserName("testUser"));

step-3:
Put user policy.

final PutUserPolicyRequest putUserPolicyRequest = new PutUserPolicyRequest()
                .withUserName("testUser").withPolicyDocument(policyDocument).withPolicyName("testPolicy");

The policyDocument I provided is:
{
	"Statement": [{
		"Action": ["s3:Get*"],
		"Effect": "Allow",
		"Resource": ["arn:aws:s3:::mybucket/myfolder/*"]
	}],
	"Version": "2012-10-17"
}

step-4:
I checked whether the identity has been created in seaweedfs by running the command s3.configure in weed shell.  I can see the identity, but its field of "actions" is an empy array. When I use the accessKey and secretKey to get the s3 object "/mybucket/myfolder/avatar.png", AccessDenied exception is received.  I don't know why my Java code above doesn't work properly to create an identity in seaweedfs. Can you please provide some help, very appreciated!

The identity I created actually by the above Java code:
    {
      "name":  "testUser",
      "credentials":  [
        {
          "accessKey":  "UKFFOIJH49X8FTIBFHSSH",
          "secretKey":  "5FktdS439208YJWROJsJQTgcfRANvjaWn2p4MZAOzW"
        }
      ],
      "actions":  [],
      "accountId":  ""
    }

The identity I expect to see:
{
      "name":  "testUser",
      "credentials":  [
        {
          "accessKey":  "UKFFOIJH49X8FTIBFHSSH",
          "secretKey":  "5FktdS439208YJWROJsJQTgcfRANvjaWn2p4MZAOzW"
        }
      ],
      "actions":  ["Read:mybucket/myfolder/*"],
      "accountId":  ""
    }

**Reply by chrislusf** [reply]
Words are not reproducible. Please share your code.

**Reply by bangbangOk** [reply]
Hi Chris, I have posted my test code below. Please help take a look, thanks.

**Reply by bangbangOk** [reply]
This is my maven config for including AWS s3 and IAM sdk.

        <dependency>
            <groupId>com.amazonaws</groupId>
            <artifactId>aws-java-sdk-s3</artifactId>
            <version>1.11.313</version>
        </dependency>

        <dependency>
            <groupId>com.amazonaws</groupId>
            <artifactId>aws-java-sdk-iam</artifactId>
            <version>1.11.313</version>
        </dependency>

**Reply by bangbangOk** [reply]
Can I have an update on this issue please?

**Reply by chrislusf** [reply]
Please try to be considerate and make it easy to reproduce. I saw other users create a repo to host the test code.

**Comment by bangbangOk** [comment]
package com.user.auth.external;

import com.amazonaws.ClientConfiguration;
import com.amazonaws.Protocol;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.AWSStaticCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.services.identitymanagement.AmazonIdentityManagement;
import com.amazonaws.services.identitymanagement.AmazonIdentityManagementClientBuilder;
import com.amazonaws.services.identitymanagement.model.*;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;

public class IAMTest {

    public static void main(String[] args) {
        createIdentity("TestUser");
    }

    public static void createIdentity(final String userName) {
        /**
         Preparation work: create a bucket, and upload a file for test:
         /buckets/mybucket/myfolder/avatar.png
         */

        // step1: Initialize IAM client using the key pair with IAM Admin permission, which is created by static configuration
        final AWSCredentials credentialsIAM = new BasicAWSCredentials("some_access_key", "some_secret_key");
        final ClientConfiguration clientConfigIAM = new ClientConfiguration();
        clientConfigIAM.setProtocol(Protocol.HTTP);
        final AmazonIdentityManagement iam = AmazonIdentityManagementClientBuilder.standard()
                .withCredentials(new AWSStaticCredentialsProvider(credentialsIAM))
                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("xx.xx.xxx.xx:8111", null))
                .withClientConfiguration(clientConfigIAM)
                .build();

        // step2: create an IAM user
        final CreateUserResult createUserResult =
                iam.createUser(new CreateUserRequest().withUserName(userName));

        // step3: create a new access key and associated it with IAM user by username
        final CreateAccessKeyResult createAccessKeyResult =
                iam.createAccessKey(new CreateAccessKeyRequest().withUserName(userName));
        final AccessKey accessKey = createAccessKeyResult.getAccessKey();

        // step3: put customized user policy
        final String policyDocument = "{\n" +
                "    \"Statement\": [{\n" +
                "        \"Action\": [\"s3:Get*\"],\n" +
                "        \"Effect\": \"Allow\",\n" +
                "        \"Resource\": [\"arn:aws:s3:::mybucket/myfolder/*\"]\n" +
                "    }],\n" +
                "    \"Version\": \"2012-10-17\"\n" +
                "}";

        final PutUserPolicyRequest putUserPolicyRequest = new PutUserPolicyRequest()
                .withUserName(userName).withPolicyDocument(policyDocument).withPolicyName(userName + "-policy");
        iam.putUserPolicy(putUserPolicyRequest);

        // step4: use the access key which is just created to initialize an S3 client
        final AWSCredentials credentialsS3 = new BasicAWSCredentials(accessKey.getAccessKeyId(), accessKey.getSecretAccessKey());
        final ClientConfiguration clientConfigS3 = new ClientConfiguration();
        clientConfigS3.setProtocol(Protocol.HTTP);
        final AmazonS3 amazonS3 = AmazonS3ClientBuilder.standard()
                .withCredentials(new AWSStaticCredentialsProvider(credentialsS3))
                .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration("xx.xx.xxx.xx:8333", null))
                .withClientConfiguration(clientConfigS3)
                .build();

        // step5: try to get the object, however an AccessDenied exception is thrown
        amazonS3.getObject("mybucket", "myfolder/avatar.png");

        // step6: check s3.configure, the identity has been created and associated with the access key, however its field "actions" is an empty array
        /*
        {
           "name":  "TestUser",
           "credentials":  [
               {
                  "accessKey":  "Z99ZIILM7NOYNSF2W0DJ9",
                  "secretKey":  "DKW1rWoLONa03CCcrxBQqp514LiiH3bQomE0MvgTJw"
               }
          ],
          "actions":  [],
          "accountId":  ""
        }
         */
    }
}

**Comment by reessing** [comment]
I just stumbled upon this issue myself. After trying to find the cause of this behavior using my almost non-existing Go knowledge, I realized that the current implementation of GetActions() in  https://github.com/seaweedfs/seaweedfs/blob/bd7068387164ebe6378aa973101364e10d250dcd/weed/iamapi/iamapi_management_handlers.go#L317 does not handle ARNs like "arn:aws:s3:::mybucket/test/\*". Even though the comment suggests otherwise, it only parses things like "arn:aws:s3:::mybucket/*" (no additional path after the bucket).

Fortunately, there already is a PR that fixes the problem (https://github.com/seaweedfs/seaweedfs/pull/3335). At least I assume it does as it includes the changes I tested to fix the problem. If this PR could be accepted, the IAM implementation of SeaweedFS would gain a very important feature.
